part_1/solution.py

```diff
@@ -0,0 +1,1327 @@
+    in_stack_depth=1,
+    pretraining=False,
+)
+
+phase2fluor_model = VSUNet(
+    architecture="UNeXt2_2D",  # 2D UNeXt2 architecture
+    model_config=phase2fluor_config.copy(),
+    loss_function=MixedLoss(l1_alpha=0.5, l2_alpha=0.0, ms_dssim_alpha=0.5),
+    schedule="WarmupCosine",
+    lr=2e-6,
+    log_batches_per_epoch=5,  # Number of samples from each batch to log to tensorboard.
+    freeze_encoder=False,
+)
+
+# %% [markdown]
+"""
+### Instantiate data module and trainer, test that we are setup to launch training.
+"""
+# %%
+source_channel = ["Phase3D"]
+target_channel = ["Nucl", "Mem"]
+# Setup the data module.
+phase2fluor_2D_data = HCSDataModule(
+    data_path,
+    architecture="UNeXt2_2D",
+    source_channel=source_channel,
+    target_channel=target_channel,
+    z_window_size=1,
+    split_ratio=0.8,
+    batch_size=BATCH_SIZE,
+    num_workers=8,
+    yx_patch_size=YX_PATCH_SIZE,
+    augmentations=augmentations,
+    normalizations=normalizations,
+)
+phase2fluor_2D_data.setup("fit")
+# fast_dev_run runs a single batch of data through the model to check for errors.
+trainer = VSTrainer(accelerator="gpu", devices=[GPU_ID], fast_dev_run=True)
+
+# trainer class takes the model and the data module as inputs.
+trainer.fit(phase2fluor_model, datamodule=phase2fluor_2D_data)
+
+
+# %% [markdown]
+# ## View model graph.
+#
+# PyTorch uses dynamic graphs under the hood.
+# The graphs are constructed on the fly.
+# This is in contrast to TensorFlow,
+# where the graph is constructed before the training loop and remains static.
+# In other words, the graph of the network can change with every forward pass.
+# Therefore, we need to supply an input tensor to construct the graph.
+# The input tensor can be a random tensor of the correct shape and type.
+# We can also supply a real image from the dataset.
+# The latter is more useful for debugging.
+
+# %% [markdown]
+# <div class="alert alert-info">
+#
+# ### Task 1.4
+# Run the next cell to generate a graph representation of the model architecture.
+# Can you recognize the UNet structure and skip connections in this graph visualization?
+# </div>
+
+# %%
+# visualize graph of phase2fluor model as image.
+model_graph_phase2fluor = torchview.draw_graph(
+    phase2fluor_model,
+    phase2fluor_2D_data.train_dataset[0]["source"][0].unsqueeze(dim=0),
+    roll=True,
+    depth=3,  # adjust depth to zoom in.
+    device="cpu",
+    # expand_nested=True,
+)
+# Print the image of the model.
+model_graph_phase2fluor.visual_graph
+
+
+# %% [markdown]
+"""
+<div class="alert alert-info">
+
+<h3> Task 1.5 </h3>
+Start training by running the following cell. Check the new logs on the tensorboard.
+</div>
+"""
+
+# %%
+# Check if GPU is available
+# You can check by typing `nvidia-smi`
+GPU_ID = 0
+
+n_samples = len(phase2fluor_2D_data.train_dataset)
+steps_per_epoch = n_samples // BATCH_SIZE  # steps per epoch.
+n_epochs = 50  # Set this to 50 or the number of epochs you want to train for.
```

Comment:
Making 50 epochs the default is *a lot* - It takes about 2min per epoch so that’s several hours of training.

Reply:
yes, I just realized... I'm going to make it 25-30 epochs.